dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
+	Claim ResourceClassClaimOptions `json:"claim,omitempty"`
+
+	// Request contains options that apply to requests in a claim.
+	Request ResourceClassRequestOptions `json:"request,omitempty"`
+
+	// DefaultRequests are individual requests for separate resources for a
+	// claim using this class. In contrast to configuration and
+	// requrirements, these requests are only used if the claim does not
+	// specify its own list of requests.
+	//
+	// If empty and the claim doesn't specify any requests either, then
+	// a single, empty request is used.
+	//
+	// +listType=atomic
+	DefaultRequests []ResourceRequest `json:"defaultRequests" protobuf:"bytes,5,name=requests"`
+}
+
+type ResourceClassClaimOptions struct {
+	// Config defines configuration parameters that apply to each claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty" protobuf:"bytes,3,opt,name=config"`
 
-	Spec DeviceClassSpec `json:"spec,omitempty"`
+	// Constraints are additional restrictions for the combination of devices that get
+	// allocated for a claim referencing this class.
+	//
+	// +optional
+	// +listType=atomic
+	Constraints []Constraint `json:"constraints,omitempty"`
 }
 
-// DeviceClassSpec provides either a selector to find other
-// classes, or the specific constraints defining this class.
-type DeviceClassSpec struct {
+type ResourceClassRequestOptions struct {
+	// Config defines configuration parameters that apply to each request in a claim using this class.
+	// They are ignored while allocating the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ConfigurationParameters `json:"config,omitempty"`
 
-	// One of selector and classCriteria must be populated.
+	// Requirements must be satisfied by devices. Applies to all devices of
+	// a claim when the claim references the class and only to the devices
+	// in a request when referenced there.
+	//
+	// +optional
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
 
-	// Selector specifies a label selector used to identify classes which
-	// may be considered equivalent to this class. In other words, this
-	// allows the adminstrator to define groups of classes which may be
-	// referred to as a single class in the claim.
+// Constraint must have one and only one field set.
+type Constraint struct {
+	// The match criteria must be satisfied before a set of devices will be used
+	// together.
 	//
 	// +optional
-	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+	Match *MatchModel `json:"match,omitempty"`
+}
 
-	// Definition specifies the criteria to determine if a device is part
-	// of this class.
-	Definition *DeviceClassDefinition `json:"definition,omitempty"`
+// ConfigurationParameters must have one and only one field set.
+type ConfigurationParameters struct {
```

Comment:
I don't understand why we have this instead of just directly referencing the CRs. This is hard to follow? Is this due to the concern about granting scheduler read access to the CRs? I think we can work through that with SIG scheduling.

Reply:
> Is this due to the concern about granting scheduler read access to the CRs?

Yes. It's not just permissions, but also performance: because there's a potentially unbounded set of types that could be referenced, the scheduler cannot use an informer cache - or at least not easily. I'm speaking from experience, I had written code for it...

> I think we can work through that with SIG scheduling.

I already discussed it with them and will not try again because they convinced me that it's not a good solution.

Reply:
I disagree. We shouldn't make our users lives significantly harder just because it's inconvenient for us to implement something. If it has major issues, sure, but I am not convinced. For example, there is an unbounded set of types in theory but in reality in any given cluster it's likely to be in the single digits or low 10s. I think we can do better here.

Reply:
Please talk to Aldo.

Reply:
BTW, I am not sure where the "significantly harder" comes from. Is more separate objects better than inlining them?

Reply:
We got approval from API reviewers to use raw extension for this particular purpose as part of the 1.30 review.

This proposal is a slight extension of the approach in 1.30: by adding the mandatory driver name, it actually becomes possible to validate as explained in the `VendorConfigurationParameters` comment.

Reply:
To be clear - some things went into 30 because we KNEW we would have to re-evaluate on a large scale, so the damage was bounded :)

Nobody loves RawExtension but it still seems "OK" for this

Reply:
Can resolve this thread